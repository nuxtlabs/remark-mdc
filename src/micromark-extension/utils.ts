import type { Effects } from 'micromark-factory-space'

// Measure the number of character codes in chunks.
// Counts tabs based on their expanded size, and CR+LF as one character.
export function sizeChunks (chunks: any[]) {
  let index = -1
  let size = 0

  while (++index < chunks.length) {
    size += typeof chunks[index] === 'string' ? chunks[index].length : 1
  }

  return size
}

export function prefixSize (events: any, type: string) {
  const tail = events[events.length - 1]
  if (!tail || tail[1].type !== type) { return 0 }
  return sizeChunks(tail[2].sliceStream(tail[1]))
}

/**
 * Calculate line indention size, line indention could be consists of multiple `linePrefix` events
 * @param events parser tokens
 * @returns line indention size
 */
export function linePrefixSize (events: any[]) {
  let size = 0
  let index = events.length - 1
  let tail = events[index]
  while (index >= 0 && tail && tail[1].type === 'linePrefix' && tail[0] === 'exit') {
    size += sizeChunks(tail[2].sliceStream(tail[1]))
    index -= 1
    tail = events[index]
  }

  return size
}

/**
 * Manage token state
 */
export const useTokenState = (tokenName: string) => {
  const token = {
    isOpen: false,
    /**
     * Enter into token, close previous open token if any
     */
    enter: (effects: Effects) => {
      const initialState = token.isOpen
      token.exit(effects)
      effects.enter(tokenName)
      token.isOpen = true

      // Revert to initial state
      return () => {
        token.isOpen = initialState
      }
    },
    /**
     * Enter into token only once, if token is already open, do nothing
     */
    enterOnce: (effects: Effects) => {
      const initialState = token.isOpen

      if (!token.isOpen) {
        effects.enter(tokenName)
        token.isOpen = true
      }

      // Revert to initial state
      return () => {
        token.isOpen = initialState
      }
    },
    /**
     * Exit from token if it is open
     */
    exit: (effects: Effects) => {
      const initialState = token.isOpen
      if (token.isOpen) {
        effects.exit(tokenName)
        token.isOpen = false
      }

      // Revert to initial state
      return () => {
        token.isOpen = initialState
      }
    }
  }
  return token
}
