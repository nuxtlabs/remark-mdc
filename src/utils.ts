export const NON_UNWRAPABLE_TYPES = [
  'componentContainerSection',
  'componentContainerDataSection',
  'containerComponent',
  'leafComponent'
]
