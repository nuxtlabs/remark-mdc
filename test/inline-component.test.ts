import { describe, expect } from 'vitest'
import { runMarkdownTests } from './utils'

describe('inline-component', () => {
  runMarkdownTests({
    empty: {
      markdown: ':component text'
    },
    text: {
      markdown: ':component[text] text'
    },
    'with-attribute': {
      markdown: ':component[text]{.class} text'
    },
    strong: {
      markdown: '**:component[text]{.class}**',
      extra (_markdown, ast, _expected) {
        expect(ast.children[0].children[0].type).toEqual('strong')
        expect(ast.children[0].children[0].children[0].type).toEqual('textComponent')
      }
    },
    underlined: {
      markdown: '**:component[text]{.class}**'
    }
  })
})
