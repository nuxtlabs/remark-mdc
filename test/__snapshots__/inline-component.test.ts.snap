// Vitest Snapshot v1

exports[`inline-component > empty 1`] = `
{
  "children": [
    {
      "children": [
        {
          "attributes": {},
          "children": [],
          "data": {
            "hName": "component",
            "hProperties": {},
          },
          "fmAttributes": {},
          "name": "component",
          "position": {
            "end": {
              "column": 11,
              "line": 1,
              "offset": 10,
            },
            "start": {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "textComponent",
        },
        {
          "position": {
            "end": {
              "column": 16,
              "line": 1,
              "offset": 15,
            },
            "start": {
              "column": 11,
              "line": 1,
              "offset": 10,
            },
          },
          "type": "text",
          "value": " text",
        },
      ],
      "position": {
        "end": {
          "column": 16,
          "line": 1,
          "offset": 15,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": {
    "end": {
      "column": 16,
      "line": 1,
      "offset": 15,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`inline-component > strong 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "attributes": {
                "class": "class",
              },
              "children": [
                {
                  "position": {
                    "end": {
                      "column": 18,
                      "line": 1,
                      "offset": 17,
                    },
                    "start": {
                      "column": 14,
                      "line": 1,
                      "offset": 13,
                    },
                  },
                  "type": "text",
                  "value": "text",
                },
              ],
              "data": {
                "hName": "component",
                "hProperties": {
                  "class": "class",
                },
              },
              "fmAttributes": {},
              "name": "component",
              "position": {
                "end": {
                  "column": 27,
                  "line": 1,
                  "offset": 26,
                },
                "start": {
                  "column": 3,
                  "line": 1,
                  "offset": 2,
                },
              },
              "type": "textComponent",
            },
          ],
          "position": {
            "end": {
              "column": 29,
              "line": 1,
              "offset": 28,
            },
            "start": {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "strong",
        },
      ],
      "position": {
        "end": {
          "column": 29,
          "line": 1,
          "offset": 28,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": {
    "end": {
      "column": 29,
      "line": 1,
      "offset": 28,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`inline-component > text 1`] = `
{
  "children": [
    {
      "children": [
        {
          "attributes": {},
          "children": [
            {
              "position": {
                "end": {
                  "column": 16,
                  "line": 1,
                  "offset": 15,
                },
                "start": {
                  "column": 12,
                  "line": 1,
                  "offset": 11,
                },
              },
              "type": "text",
              "value": "text",
            },
          ],
          "data": {
            "hName": "component",
            "hProperties": {},
          },
          "fmAttributes": {},
          "name": "component",
          "position": {
            "end": {
              "column": 17,
              "line": 1,
              "offset": 16,
            },
            "start": {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "textComponent",
        },
        {
          "position": {
            "end": {
              "column": 22,
              "line": 1,
              "offset": 21,
            },
            "start": {
              "column": 17,
              "line": 1,
              "offset": 16,
            },
          },
          "type": "text",
          "value": " text",
        },
      ],
      "position": {
        "end": {
          "column": 22,
          "line": 1,
          "offset": 21,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": {
    "end": {
      "column": 22,
      "line": 1,
      "offset": 21,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`inline-component > underlined 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "attributes": {
                "class": "class",
              },
              "children": [
                {
                  "position": {
                    "end": {
                      "column": 18,
                      "line": 1,
                      "offset": 17,
                    },
                    "start": {
                      "column": 14,
                      "line": 1,
                      "offset": 13,
                    },
                  },
                  "type": "text",
                  "value": "text",
                },
              ],
              "data": {
                "hName": "component",
                "hProperties": {
                  "class": "class",
                },
              },
              "fmAttributes": {},
              "name": "component",
              "position": {
                "end": {
                  "column": 27,
                  "line": 1,
                  "offset": 26,
                },
                "start": {
                  "column": 3,
                  "line": 1,
                  "offset": 2,
                },
              },
              "type": "textComponent",
            },
          ],
          "position": {
            "end": {
              "column": 29,
              "line": 1,
              "offset": 28,
            },
            "start": {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "strong",
        },
      ],
      "position": {
        "end": {
          "column": 29,
          "line": 1,
          "offset": 28,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": {
    "end": {
      "column": 29,
      "line": 1,
      "offset": 28,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`inline-component > with-attribute 1`] = `
{
  "children": [
    {
      "children": [
        {
          "attributes": {
            "class": "class",
          },
          "children": [
            {
              "position": {
                "end": {
                  "column": 16,
                  "line": 1,
                  "offset": 15,
                },
                "start": {
                  "column": 12,
                  "line": 1,
                  "offset": 11,
                },
              },
              "type": "text",
              "value": "text",
            },
          ],
          "data": {
            "hName": "component",
            "hProperties": {
              "class": "class",
            },
          },
          "fmAttributes": {},
          "name": "component",
          "position": {
            "end": {
              "column": 25,
              "line": 1,
              "offset": 24,
            },
            "start": {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "textComponent",
        },
        {
          "position": {
            "end": {
              "column": 30,
              "line": 1,
              "offset": 29,
            },
            "start": {
              "column": 25,
              "line": 1,
              "offset": 24,
            },
          },
          "type": "text",
          "value": " text",
        },
      ],
      "position": {
        "end": {
          "column": 30,
          "line": 1,
          "offset": 29,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": {
    "end": {
      "column": 30,
      "line": 1,
      "offset": 29,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;
